### PR TITLE
remove fields on unsupported plugin version (WP-814)

### PR DESCRIPTION
**Pull request by vsolovei-smartling**
Root cause of issue was unsupported plugin version. Research took some time because I tried reproducing the issue. To avoid this in the future, I propose to remove content that is obviously unsuitable for translation in cases we detect an unsupported version of a plugin

### CODE REVIEW

File: inc/Smartling/ContentTypes/ExternalContentManager.php [hunk 1]
```diff
@@ -58,7 +72,7 @@ public function getExternalRelations(string $contentType, int $id): array
     {
         $result = [];
         foreach ($this->handlers as $handler) {
-            if ($handler->canHandle($contentType, $id)) {
+            if ($handler->getSupportLevel($contentType, $id)) {
```
Comment (PavelLoparev) [inline inc/Smartling/ContentTypes/ExternalContentManager.php:75]: Same here (about always true condition)

Comment (PavelLoparev) [reply inc/Smartling/ContentTypes/ExternalContentManager.php:75]: `canHandle()` returned bool indeed.

File: inc/Smartling/ContentTypes/ContentTypePluggableInterface.php [hunk 2]
```diff
@@ -2,11 +2,17 @@
...
 use Smartling\Submissions\SubmissionEntity;
 
 interface ContentTypePluggableInterface
 {
-    public function canHandle(string $contentType, ?int $contentId = null): bool;
+    public const NOT_SUPPORTED = 'not_supported';
+    public const SUPPORTED = 'supported';
+    public const VERSION_NOT_SUPPORTED = 'version_not_supported';
+
+    #[ExpectedValues(valuesFromClass: self::class)]
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ContentTypePluggableInterface.php:14]: Phpstorm only

File: inc/Smartling/ContentTypes/ExternalContentElementor.php [hunk 3]
```diff
@@ -146,11 +146,12 @@ public function alterContentFieldsForUpload(array $source): array
...
-    public function canHandle(string $contentType, ?int $contentId = null): bool
+    public function getSupportLevel(string $contentType, ?int $contentId = null): string
     {
-        return parent::canHandle($contentType, $contentId) &&
-            $this->contentTypeHelper->isPost($contentType) &&
-            $this->getDataFromPostMeta($contentId) !== '';
+        if ($this->contentTypeHelper->isPost($contentType) && $this->getDataFromPostMeta($contentId) !== '') {
+            return parent::getSupportLevel($contentType, $contentId);
+        }
+        return self::NOT_SUPPORTED;
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ExternalContentElementor.php:154]: There are two classes that check for not empty metadata, but other implementations vary

File: inc/Smartling/ContentTypes/ExternalContentManager.php [hunk 4]
```diff
@@ -23,21 +23,35 @@ public function addHandler(ContentTypePluggableInterface $handler): void
...
+                        try {
+                            $source = $handler->alterContentFieldsForUpload($source);
+                        } catch (\Throwable $e) {
+                            $this->getLogger()->notice('HandlerName="' . $handler->getPluginId() . '" got exception while trying to alter content fields: ' . $e->getMessage());
+                        }
+                    }
+                    break;
+                case ContentTypePluggableInterface::VERSION_NOT_SUPPORTED:
+                    if ($handler instanceof ContentTypeModifyingInterface) {
+                        $this->getLogger()->warning("Detected not supported version for {$handler->getPluginId()}, will clear known problematic fields");
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ExternalContentManager.php:45]: Removed from generated xml

File: inc/Smartling/ContentTypes/ExternalContentGravityForms.php [hunk 5]
```diff
@@ -43,7 +43,7 @@ public function __construct(
         $this->gutenbergBlockHelper = $gutenbergBlockHelper;
         $this->handler = $handler;
         $this->siteHelper = $siteHelper;
-        if (parent::canHandle(self::CONTENT_TYPE)) {
+        if (parent::getSupportLevel(self::CONTENT_TYPE)) {
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ExternalContentGravityForms.php:46]: It does, thanks, fixed

File: inc/Smartling/ContentTypes/ExternalContentElementor.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (PavelLoparev) [reply inc/Smartling/ContentTypes/ExternalContentElementor.php:154]: Sounds like sub parent is needed for the classes that check for not empty metadata. Just thinking, no action points needed I think, sounds like more about refactoring.

File: inc/Smartling/ContentTypes/ExternalContentManager.php [hunk 6]
```diff
@@ -84,7 +98,7 @@ public function getHandlers(): array
     public function setExternalContent(array $original, array $translation, SubmissionEntity $submission): array
     {
         foreach ($this->handlers as $handler) {
-            if ($handler->canHandle($submission->getContentType(), $submission->getSourceId())) {
+            if ($handler->getSupportLevel($submission->getContentType(), $submission->getSourceId())) {
```
Comment (PavelLoparev) [inline inc/Smartling/ContentTypes/ExternalContentManager.php:101]: Please make sure you updated all the places with such condition

Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ExternalContentManager.php]: Done!

**Review by PavelLoparev — APPROVED** [review]

**Pull request merged**